Botonic Plugin Google Analytics

asastre opened the pull request:
## Description
New plugin to track user interaction or bot's behaviour to Google Analytics.

## Context
Tracking information to Google Analytics is very usual nowadays and for some companies there is the need to track the user interaction with the bot to Google Analytics.

## Approach taken / Explain the design
This plugin provides a basic tracking with its method `track` and it has to be done manually, there is no default tracking through the `pre` or `post` methods.

## To documentate / Usage example
See [Readme.md](https://github.com/hubtype/botonic/blob/8be257fd2b8b7b2701b5d420bd8d89a0c9dbb1cb/packages/botonic-plugin-google-analytics/README.md) file for documentation/examples.

## Testing

The pull request...

- [ ] has unit tests
- [ ] has integration tests
- [ ] doesn't need tests because... no tests needed

dpinol commented on packages/botonic-plugin-google-analytics/README.md:
```diff
@@ -0,0 +1,49 @@
+    const ga = plugins.google_analytics
+    
+    const eventName = 'botEvent'
+    const eventData = {
```
naming consistency.
In the plugin the field is called eventFields

asastre replied:
@dpinol corrected

dpinol commented on packages/botonic-plugin-google-analytics/package.json:
```diff
@@ -0,0 +1,17 @@
+    "cloc": "../../scripts/qa/cloc-package.sh .",
+    "prepare": "node ../../preinstall.js",
+    "lint": "npm run lint_core -- --fix",
+    "lint_ci": "npm run lint_core",
```
clone .github/workflows/botonic-plugin-dashbot-tests.yml to create a github actions CI check

dpinol replied:
Add a new hook to .pre-commit-config.yaml so that lint is executed when we do git commit

dpinol commented on packages/botonic-plugin-google-analytics/package.json:
```diff
@@ -0,0 +1,17 @@
+    "lint_core": "../../node_modules/.bin/eslint_d --cache --quiet 'src/**/*.js'"
+  },
+  "dependencies": {
+    "@analytics/google-analytics": "^0.4.0",
```
since we don't have integration tests, I would be conservatieve and use "~" for both dependencies

asastre replied:
@dpinol done

dpinol commented on packages/botonic-plugin-google-analytics/package.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "@botonic/plugin-google-analytics",
+  "version": "0.14.0",
```
only use 0.14.0 if you're going to publish it after merge. Otherwise, 0.14.0-rc1

vanbasten17 replied:
I think it will be published as it's needed for a project, so it's okay that it is `0.14.0`.

dpinol commented on packages/botonic-plugin-google-analytics/README.md:
```diff
@@ -0,0 +1,49 @@
+    
+    const eventName = 'botEvent'
+    const eventData = {
+      category: 'chatbot',
```
please document what eventData/eventFields can/should contain:
* only the fields in the example or any fields?
* only strings and numbers, also nested objects, also class instances?

vanbasten17 replied:
@dpinol is it possible to add an `index.d.ts` for this project with definitions, or it has to be strictly a ts project?

asastre replied:
@dpinol @vanbasten17 I enhanced the documentation and I added descriptions of the eventFields argument.

vanbasten17 replied:
looking nice! 👌

dpinol replied:
@vanbasten17 good point, we have index.d.ts for core and react, right? Can you help Sastre we this?

asastre replied:
@dpinol great! I'll have a look at it with @vanbasten17 in next PR

vanbasten17 commented on packages/botonic-plugin-google-analytics/README.md:
```diff
@@ -0,0 +1,49 @@
+## Use
+
+The tracking needs to be done manually, enabling this plugin doesn't track any user interaction or bot's behaviour by default.
+You can do the tracking inside the `botonicInit` method, `render` method or even inside the [Webchat listeners](https://docs.botonic.io/concepts/webchat#webchat-listeners)
```
i'm not pretty sure about the usage inside the listeners itself (I've never used a plugin in this way), have you tried it @asastre ?

asastre replied:
@dpinol maybe is better to remove the `async` `await` from the `track` method because it doesn't make sense to stop the execution of the bot while doing the tracking and hence it could be called anywhere (`botonicInit`, `render` or webchat listeners)

asastre replied:
@vanbasten17 I've tryed to call it inside the `onInit` method and it works. I'll use it here to track the impressions of the chatbot for example.

dpinol replied:
> @dpinol maybe is better to remove the `async` `await` from the `track` method because it doesn't make sense to stop the execution of the bot while doing the tracking and hence it could be called anywhere (`botonicInit`, `render` or webchat listeners)

You should not remove async, await. Or if you do, you should return the promise. Otherwise:
* you don't let the bot wait for the response in case they want to do it
* in massimo's bot where we track to dynamo, if you send the track when the lambda instance is restarted (it happens every few hours), then if you don't have the await, the track is never sent. I guess here would happen the same

asastre commented:
@dpinol @vanbasten17 after doing some tests with the Google Analytics tracking I did a refactor of the `track` method to match correctly the fields of eventFields with the GA fields.

I've removed the `async`/`await` of `track` method because, as I said in a comment before, it doesn't make sense to stop the execution of the bot while doing the tracking.

I've also added more and better documentation of the `track` method.

ericmarcos commented on packages/botonic-plugin-google-analytics/src/index.js:
```diff
@@ -0,0 +1,40 @@
+          trackingId: trackingId,
+        }),
+      ],
+    })
```
You'll also need to call the `identify` method so that GA knows what user is sending the events. Something like
```
this.analytics.identify(session.user.id, {
  firstName: session.user.name,
  channel: session.user.provider,
  channel_id: session.user.provider_id
})
```
I'm not sure if you can set any params you want...
Also, I'm not sure if you can do this in the constructor or you'll nedd to do it in every `track` call.

ericmarcos commented:
> @dpinol @vanbasten17 after doing some tests with the Google Analytics tracking I did a refactor of the `track` method to match correctly the fields of eventFields with the GA fields.
> 
> I've removed the `async`/`await` of `track` method because, as I said in a comment before, it doesn't make sense to stop the execution of the bot while doing the tracking.
> 
> I've also added more and better documentation of the `track` method.

@asastre please leave the `async` and state in the docs that you must `await` this call. As we're running this in Lambda, the main "thread" can finish and the process terminated leaving async calls that are not awaited terminated before they end the execution.

asastre commented on packages/botonic-plugin-google-analytics/README.md:
```diff
@@ -0,0 +1,51 @@
+    id: 'google_analytics',
+    resolve: require('@botonic/plugin-google-analytics'),
+    options: {
+      botName: 'YOUR_BOT_NAME',
```
@ericmarcos yes, I'll remove it :+1:

asastre commented:
@dpinol @vanbasten17 @ericmarcos changes done. I restored the `async`/`await` in `track` method, as you said, in production the plugin will be executed in lambda and it needs to be sure it will execute.

dpinol commented:
@asastre review the prettier errors in CI checks. More important, your pre-commit should not have let you comit this file. Please verify that you did "pre-commit install" on your botonic checkout
/home/runner/work/botonic/botonic/packages/botonic-plugin-google-analytics/src/index.js
##[error]  62:35  error  Replace `this.userId({·session·}),·this.userTraits({·session·})` with `⏎······this.userId({·session·}),⏎······this.userTraits({·session·})⏎····`  prettier/prettier

dpinol commented on packages/botonic-plugin-google-analytics/src/index.js:
```diff
@@ -2,39 +2,64 @@ import Analytics from 'analytics'
+  /**
+   * @param {Object} options - Options for the plugin
+   * @param {string} options.trackingId - Tracking ID for Google Analytics.
+   * @param {function({session: object}): string} [options.userId] - Method that returns a unique user ID as string.
```
maybe it's more clear to call it getUserId & getUserTraits?

asastre commented:
@dpinol @ericmarcos I've made some changes/enhancements:

- Refactor name methods `defaultUserId` and `defaultUserTraits` to `getUserId` and `getUserTraits`.
- Refactor the option `trackManually` to `automaticTracking` to clarify its usage. If this option is not set, its default value will be `true`, enabling the automatic tracking on each user interaction.
- Added a new option: `eventFields` to enable to send custom event fields to the automatic tracking done by the plugin (only used if `automaticTracking` is not set or is set to `true`)

dpinol commented on packages/botonic-plugin-google-analytics/src/index.js:
```diff
@@ -10,8 +10,8 @@ export default class BotonicPluginGoogleAnalytics {
   constructor(options) {
-    this.userId = options.userId ?? this.defaultUserId
-    this.userTraits = options.userTraits ?? this.defaultUserTraits
+    this.userId = options.userId ?? this.getUserId
```
sorry I meant to rename options.userId to options.getUserId so that it's clear for the developer that it's a function. what do you think?
about the defaultUserId, defaultUserTraits functions, since the developers don't need to interact with them, their name is not so important. defaultGetUserId or getUserIdFromSession sound good to me

asastre replied:
Oh! ok, now I understand and I do like the rename, it's clearer. Done! Thanks!

dpinol commented on packages/botonic-plugin-google-analytics/src/index.js:
```diff
@@ -0,0 +1,75 @@
+   * @param {function({session: object}): object} [options.userTraits] - Method that returns the user traits as object.
+   * @param {boolean} [options.automaticTracking] - If set to false, no automatic tracking will be done in post method.
+   * @param {function({session: object, input: object, lastRoutePath: string}): object} [options.eventFields] - Method
+   *    that returns the eventFields to track as object (used only if automaticTracking is set to false).
```
you meant "if automaticTracking is set to TRUE", right?

asastre replied:
yes! Good eyes :eyes: ! Corrected.

asastre commented:
> can you comment on why you removed the default getUserId?

I removed it because at first we were assuming that `clientId` and `userId` were the same for Google Analytics, but they are [different concepts](https://support.google.com/analytics/answer/6205850?hl=en#clientid-userid) and `userId` is meant to represent a logged user, so the plugin can't add one by default, the user must provide one if they want to track it.

asastre commented:
@dpinol @ericmarcos The plugin wasn't working properly when tracking on the server side loosing events randomly and never reaching Google Analytics.
After trying various tracking packages (`@analytics/google-analytics`, `universal-ga`) and even [proposing changes](https://github.com/DavidWells/analytics/pull/64) to one of them so the `clientId` parameter could be sent in the server side tracking (feature needed to keep the tracking flow for both client and server side), the tracking on the server side didn't work as expected, probably because it's executed in Lambda.
The new aproach (which is working fine) is to make direct requests to Google Analytics (using `axios`) sending the parameters needed each time.

asastre commented on packages/botonic-plugin-google-analytics/src/index.js:
```diff
@@ -0,0 +1,76 @@
+      ev: eventFields.value,
+    }
+    params.cid =
+      this.getGaClientId && this.getGaClientId({ session })
```
Totally agree, changed!

asastre commented on packages/botonic-plugin-google-analytics/src/index.js:
```diff
@@ -0,0 +1,76 @@
+      this.getGaClientId && this.getGaClientId({ session })
+        ? (params.cid = this.getGaClientId({ session }))
+        : uuidv4()
+    if (this.getUserId && this.getUserId({ session }))
```
This case is similar but it's not the same because we only want to add the option `params.uid` if `getUserId` is set and returns something. 
What I've changed here is not to call 2 times `this.getUserId` as you well said in other comment.
